File: android/app/src/main/java/cz/qrplatba/gateway/BankGateway.kt
package cz.qrplatba.gateway

import cz.qrplatba.domain.Money
import java.math.BigDecimal

/** An incoming bank transaction. externalId is used for idempotence. */
data class BankTransaction(
    val externalId: String,
    val amount: Money,
    val currency: String,
    val vs: String?,
    val receivedAt: Long,
)

/** Raised by a gateway when the bank cannot be reached (drives UNKNOWN handling). */
class BankUnavailableException(message: String) : Exception(message)

/**
 * The single seam between matching logic and any bank. Swapping
 * SimulatorGateway -> FioGateway must not change matching, session states, or the API.
 */
interface BankGateway {
    /** New incoming transactions since the last checkpoint. Throws [BankUnavailableException] if unreachable. */
    fun fetchNewTransactions(): List<BankTransaction>

    /** Connectivity probe — drives UNKNOWN / "cannot verify" handling. */
    fun isAvailable(): Boolean
}

File: server/src/gateway/BankGateway.ts
import type { Money } from "../domain/money.js";

export interface BankTransaction {
  externalId: string; // bank's transaction id — used for idempotence
  amount: Money;
  currency: string; // "CZK"
  vs: string | null; // variable symbol
  receivedAt: Date;
}

/**
 * The single seam between our matching logic and any bank.
 * Swapping SimulatorGateway -> FioGateway must not change matching, session
 * states, or the API (AC-0.2).
 */
export interface BankGateway {
  /** New incoming transactions since the last checkpoint. */
  fetchNewTransactions(): Promise<BankTransaction[]>;
  /** Connectivity probe — drives UNKNOWN / "cannot verify" handling. */
  isAvailable(): Promise<boolean>;
}
